kernel/sched.c

```diff
@@ -1347,3 +1378,78 @@ int k_thread_cpu_mask_disable(k_tid_t thread, int cpu)
 }
 
 #endif /* CONFIG_SCHED_CPU_MASK */
+
+int z_impl_k_thread_join(struct k_thread *thread, s32_t timeout)
+{
+	k_spinlock_key_t key;
+	int ret;
+
+	__ASSERT(((arch_is_in_isr() == false) || (timeout == K_NO_WAIT)), "");
+
+	key = k_spin_lock(&sched_spinlock);
+
+	if ((thread->base.thread_state & _THREAD_DEAD) != 0) {
+		ret = 0;
+		goto out;
+	}
+
+	if (timeout == K_NO_WAIT) {
+		ret = -EBUSY;
+		goto out;
+	}
+
+	if (thread->base.pended_on == &_current->base.join_waiters) {
```

Comment:
What's the circumstance where this would be true?  When pended_on is non-NULL, a thread isn't supposed to be running.

Reply:
This is the case where we are attempting to join a thread, who has previously gone to sleep trying to join on us -- they are asleep on this thread's join_waiters queue.

I added this because it seemed a simple check to make and pthread_join() does it as well.

However, it seems I have no test coverage of this scenario, will add.

Reply:
test coverage of this now added

Reply:
Do we care?  That's only one of an infinite space of dependency cycles (i.e. it can't detect A joins B joins C joins A, or A joins B waiting for A to do something, etc...).    Other IPC mechanisms don't check for this kind of thing.

Reply:
> Do we care?

Fair point. I put the check pretty much because pthread does it and it was cheap. I'm not opposed to removing.